test/export/InstanceExporter.test.ts

```diff
@@ -333,5 +341,100 @@ describe('InstanceExporter', () => {
+        null,
+        { url: 'type', valueCoding: { system: 'foo', version: '1.2.3' } }
+      ]);
+    });
+
+    it('should fix sliced elements in an array that are fixed out of order', () => {
+      const fooRule = new FixedValueRule('extension[type][1].valueCoding.system');
+      fooRule.fixedValue = 'foo';
+      patientProfInstance.rules.push(fooRule);
+      const barRule = new FixedValueRule('extension[type][0].valueCoding.system');
+      barRule.fixedValue = 'bar';
+      patientProfInstance.rules.push(barRule);
+      const exported = exporter.exportInstance(patientProfInstance);
+      expect(exported.extension).toEqual([
+        { url: 'type', valueCoding: { system: 'bar' } },
+        { url: 'type', valueCoding: { system: 'foo' } }
+      ]);
+    });
+
+    it('should fix sliced elements in an array and null fill empty values', () => {
+      const fooRule = new FixedValueRule('extension[type][1].valueCoding.system');
+      fooRule.fixedValue = 'foo';
+      patientProfInstance.rules.push(fooRule);
+      const exported = exporter.exportInstance(patientProfInstance);
+      expect(exported.extension).toEqual([null, { url: 'type', valueCoding: { system: 'foo' } }]);
+    });
+
+    it('should fix mixed sliced elements in an array', () => {
+      const fooRule = new FixedValueRule('extension[type][0].valueCoding.system');
+      fooRule.fixedValue = 'foo';
+      patientProfInstance.rules.push(fooRule);
+      const bazRule = new FixedValueRule('extension[level].valueCoding.system');
+      bazRule.fixedValue = 'baz';
+      patientProfInstance.rules.push(bazRule);
+      const barRule = new FixedValueRule('extension[type][1].valueCoding.system');
+      barRule.fixedValue = 'bar';
+      patientProfInstance.rules.push(barRule);
+      const exported = exporter.exportInstance(patientProfInstance);
+      expect(exported.extension).toEqual([
+        { url: 'type', valueCoding: { system: 'foo' } },
+        { url: 'level', valueCoding: { system: 'baz' } },
+        { url: 'type', valueCoding: { system: 'bar' } }
+      ]);
+    });
+
+    it('should fix mixed sliced elements in an array out of order', () => {
+      const fooRule = new FixedValueRule('extension[type][1].valueCoding.system');
+      fooRule.fixedValue = 'foo';
+      patientProfInstance.rules.push(fooRule);
+      const bazRule = new FixedValueRule('extension[level].valueCoding.system');
+      bazRule.fixedValue = 'baz';
+      patientProfInstance.rules.push(bazRule);
+      const barRule = new FixedValueRule('extension[type][0].valueCoding.system');
+      barRule.fixedValue = 'bar';
+      patientProfInstance.rules.push(barRule);
+      const exported = exporter.exportInstance(patientProfInstance);
+      expect(exported.extension).toEqual([
+        { url: 'type', valueCoding: { system: 'bar' } },
+        { url: 'type', valueCoding: { system: 'foo' } },
+        { url: 'level', valueCoding: { system: 'baz' } }
+      ]);
+    });
```

Comment:
This is good, but I think we need a test for the thing I asked about above... essentially adding a new element to the end of a slice in the middle of the array.  Something like:

```ts
    it('should fix mixed sliced elements in an array out of order', () => {
      const barRule = new FixedValueRule('extension[type][0].valueCoding.system');
      barRule.fixedValue = 'bar';
      patientProfInstance.rules.push(barRule);
      const bazRule = new FixedValueRule('extension[level].valueCoding.system');
      bazRule.fixedValue = 'baz';
      patientProfInstance.rules.push(bazRule);
      const fooRule = new FixedValueRule('extension[type][1].valueCoding.system');
      fooRule.fixedValue = 'foo';
      patientProfInstance.rules.push(fooRule);
      const exported = exporter.exportInstance(patientProfInstance);
      expect(exported.extension).toEqual([
        { url: 'type', valueCoding: { system: 'bar' } },
        { url: 'type', valueCoding: { system: 'foo' } },
        { url: 'level', valueCoding: { system: 'baz' } }
      ]);
    });
```

Reply:
I think that the test starting on line `404, "should fix mixed sliced elements in an array" tests this situation if I'm not missing something. As the test shows, the `level` slice does not get overwritten, but the second `type` slice gets added to the end of the array, not in the middle:
```
expect(exported.extension).toEqual([
        { url: 'type', valueCoding: { system: 'foo' } },
        { url: 'level', valueCoding: { system: 'baz' } },
        { url: 'type', valueCoding: { system: 'bar' } }
      ]);
```
Is that an issue? Are slices of the same type supposed to be adjacent?

Reply:
Huh. Yeah, line 404 looks a lot like the test I just suggested.  Obviously I must have been very sorry while doing this review.  I'm sorry...  TBH, I'm not sure if slice parts need to be grouped together.  We might want to look at the slicing doc to be sure.

One thing we're not paying attention to here is if the discriminator says slices are ordered.  If they're ordered then we need to force the author to declare the slice values in order or we need to re-order them for the author.

The other thing is open/closed.  If slicing is closed then the author should not be allowed to add any elements that don't fit in a slice.  That one might be a bit harder to enforce though.  It would only work if the author used the slice naming syntax...  Should we force that for closed slicings?

Reply:
I don't think the slicing doc says anything explicitly, so I guess we just want to go with whatever behavior is expected. I guess it probably would make sense to have slice parts grouped together, since the definition of slicing does say "sub-list". To me that implies all the elements of the list are together, and at the very least it seems to be a best practice used in all the examples I see.

I will look at adding checking of the ordering, I'd prefer to throw an error at the author, not to re-order, since that could lead to confusion in how the FSH correlates to the SD being created.

We would have to use the explicit slice naming syntax because otherwise we would have to try to check which slice something fits right? And that would involve using the discriminator and checking fixed values and overall be pretty difficult, correct? And when you say slice naming syntax you mean the `Slice` keyword right? Assuming the answer to all those questions is yes or mostly yes, then yeah I think it makes sense to enforce that for closed slicings. But I don't know if we want to do that yet, since the `Slice` keyword doesn't exist yet, we would effectively be shutting down and closed slicings in the meantime. So until the `Slice` keyword would it make sense to just allow anything for closed?